service/src/main/java/bio/terra/workspace/service/resource/controlled/flight/clone/azure/database/CopyControlledAzureDatabaseDefinitionStep.java

```diff
@@ -0,0 +1,223 @@
+  private final ControlledAzureDatabaseResource sourceDatabase;
+  private final ControlledResourceService controlledResourceService;
+  private final CloningInstructions resolvedCloningInstructions;
+  private final WsmResourceService wsmResourceService;
+
+  public CopyControlledAzureDatabaseDefinitionStep(
+      SamService samService,
+      AuthenticatedUserRequest userRequest,
+      ControlledAzureDatabaseResource sourceDatabase,
+      ControlledResourceService controlledResourceService,
+      CloningInstructions resolvedCloningInstructions,
+      WsmResourceService wsmResourceService) {
+    this.samService = samService;
+    this.userRequest = userRequest;
+    this.sourceDatabase = sourceDatabase;
+    this.controlledResourceService = controlledResourceService;
+    this.resolvedCloningInstructions = resolvedCloningInstructions;
+    this.wsmResourceService = wsmResourceService;
+  }
+
+  @Override
+  public StepResult doStep(FlightContext flightContext)
+      throws InterruptedException, RetryException {
+    var inputParameters = flightContext.getInputParameters();
+    var workingMap = flightContext.getWorkingMap();
+
+    // get the inputs from the flight context
+    var destinationResourceName =
+        getInputParameterOrWorkingValue(
+            flightContext,
+            WorkspaceFlightMapKeys.ResourceKeys.RESOURCE_NAME,
+            WorkspaceFlightMapKeys.ResourceKeys.PREVIOUS_RESOURCE_NAME,
+            String.class);
+    var description =
+        getInputParameterOrWorkingValue(
+            flightContext,
+            WorkspaceFlightMapKeys.ResourceKeys.RESOURCE_DESCRIPTION,
+            WorkspaceFlightMapKeys.ResourceKeys.PREVIOUS_RESOURCE_DESCRIPTION,
+            String.class);
+    var sourceWorkspaceId = sourceDatabase.getWorkspaceId();
+    var destinationWorkspaceId =
+        getRequired(inputParameters, ControlledResourceKeys.DESTINATION_WORKSPACE_ID, UUID.class);
+    var destinationResourceId =
+        getRequired(inputParameters, ControlledResourceKeys.DESTINATION_RESOURCE_ID, UUID.class);
+
+    var sourceIdentity =
+        wsmResourceService
+            .enumerateResources(
+                sourceWorkspaceId,
+                WsmResourceFamily.AZURE_MANAGED_IDENTITY,
+                StewardshipType.CONTROLLED,
+                0,
+                100)
+            .stream()
+            .filter(
+                (wsmResource) ->
+                    wsmResource.getName().equals(sourceDatabase.getDatabaseOwner())
+                        || wsmResource
+                            .getResourceId()
+                            .toString()
+                            .equals(sourceDatabase.getDatabaseOwner()))
+            .findFirst()
+            .orElseThrow(
+                () ->
+                    new IllegalStateException(
+                        "Could not find managed identity that owns database %s"
+                            .formatted(sourceDatabase.getResourceId())));
```

Comment:
why not use [getResourceByName](https://github.com/DataBiosphere/terra-workspace-manager/blob/ad7ebaffadb34f5d87291de438ecaec34e8799c5/service/src/main/java/bio/terra/workspace/db/ResourceDao.java#L567), below too. Shouldn't the destination managed identity WSM resource have the same name as the source?

Reply:
The database ownership seems to refer to either the managed identity resource name or resource id, which is why the custom check is needed for the sourceIdentity. For destination, I suppose we could assume that it will have the same name but this is only necessarily true in workspace clones, and would not hold for eventual individual identity/database clones where names can be specified different from the original, whereas they would still have the correct resource lineage.

From the swagger:
> Resource id or name of an azure managed identity in this workspace to be the owner of the database, required for shared databases. For private databases, the owner will be determined from privateResourceUser.

Reply:
only the name of the resource is stored in the database. Any ids that may be in the database currently are old and can be ignored. The api accepts id for backwards compatibility but translates to name before storing.